### PR TITLE
Add support for uneven batch split across data-parallel configurations

### PR DESCRIPTION
**Pull request by rdraskicTT**
### Ticket
https://github.com/tenstorrent/tt-metal/issues/23042

### Problem description
Running vllm benchmark hits an issue where `unpadded_batch` is not a proper multiple of `data_parallel` parameter so we need a better handling of edge cases. It's supposed to enable [this PR in vllm](https://github.com/tenstorrent/vllm/pull/97).

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [x] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/15346865868) CI passes
- [ ] [Blackhole Post commit](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml) CI with demo tests passes (if applicable)
- [ ] [Model regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-models.yaml) CI passes (if applicable)
- [ ] [Device performance regression](https://github.com/tenstorrent/tt-metal/actions/workflows/perf-device-models.yaml) CI passes (if applicable)
- [ ] (For models and ops writers) [Single-card demo tests](https://github.com/tenstorrent/tt-metal/actions/workflows/single-card-demo-tests.yaml) CI passes (if applicable) See [recommended dev flow](https://github.com/tenstorrent/tt-metal/blob/main/models/MODEL_ADD.md#a-recommended-dev-flow-on-github-for-adding-new-models).
- [ ] New/Existing tests provide coverage for changes

### CODE REVIEW

File: models/tt_transformers/tt/generator.py [hunk 1]
```diff
@@ -599,14 +599,18 @@ def decode_forward(
             return tt_logits
 
     # Note: This function is called by vLLM
-    def read_decode_output(self, tt_out, unpadded_batch, is_tokens=False):
+    def read_decode_output(self, tt_out, unpadded_batch, padded_batch=None, is_tokens=False):
         """
         Input is ttnn device tensor of logits if is_tokens=False, otherwise tokens. Output is the corresponding torch tensor.
         """
-        batch_per_dp = unpadded_batch // self.data_parallel
```
Comment (ppetrovicTT) [inline models/tt_transformers/tt/generator.py:606]: i wonder if it would be enough (for this whole PR) to just ceil this?
```
batch_per_dp = (unpadded_batch + self.data_parallel -1) // self.data_parallel
```

`tt_out` should already be padded, so in case of just 1 query, only one batch should be of interest, but the rest should be empty.. but it should still work, no?

Comment (rdraskicTT) [reply models/tt_transformers/tt/generator.py:606]: I think that what you suggest works only for `unpadded_batch=1`. This formula is required because when we run `benchmarks/benchmark_serving.py` and setting `--num-prompts` what happens is that the sequence size is calculated based on `--max_num_seqs` from the other PR as well. For example, let's say that `--max_num_seqs=32` and `--num-prompts=1000`. The runs will be split into many that send 32 prompts each but the last run will receive 8 prompts. In that case `unpadded_batch=8`, however it's padded to batch 32 and then this is being split across dp instances. Let's say `data_parallel=4`, then all 8 prompts end up on the first dp instance. What you suggest makes sense if they are being split equally among the dp instances but that's not how it works currently. Maybe we can get into whether it's more performant to split it differently but I think it's a topic for a different PR.

**Review by skhorasganiTT — CHANGES_REQUESTED** [review]
Synced offline with Radoica about this. The padded_batch parameter should not be needed. `_decode_forward_no_trace_text` should be returning a list (DP elements) with each element containing the same batch size corresponding to the inputs. That currently  might not be the case and a slice may need to be added to that function if it's padding the batch to tile size 32. Once that is fixed, `read_decode_output` should read the total batch_per_dp for each DP model, and at the very end after converting to torch, it should slice to the unpadded_batch which corresponds to the total DP batch.

File: models/tt_transformers/tt/generator.py [hunk 2]
```diff
@@ -599,14 +599,18 @@ def decode_forward(
             return tt_logits
 
     # Note: This function is called by vLLM
-    def read_decode_output(self, tt_out, unpadded_batch, is_tokens=False):
+    def read_decode_output(self, tt_out, unpadded_batch, padded_batch=None, is_tokens=False):
```
Comment (ipotkonjak-tt) [inline models/tt_transformers/tt/generator.py:602]: `padded_batch` seems unnecessary. `unpadded_batch` and `self.data_parallel` should be enough to get the ceil/floor of batch per DP group and slice out padded values.

**Comment by skhorasganiTT** [comment]
Should this PR be closed now that there's #23129?

**Comment by rdraskicTT** [comment]
> Should this PR be closed now that there's #23129?

Yes, I'll reference this one over there.
